Technical/Logging/The-Tazama-Logging-Framework.md

```diff
@@ -1 +1,405 @@
+
+### Implementation
+The [event-sidecar] and [lumberjack] are optional components in implementation. Perhaps hardware is limited or there may be some situations whereby running more microservices is not ideal, Tazama will still have a functional logging implementation if the [event-sidecar] and [lumberjack] are excluded from deployment. As mentioned before, you can opt in to using the [event-sidecar] and [lumberjack] for logging by specifying the sidecar host when initialising the logger service.
+
+#### Configuration
+
+##### With Event-Sidecar and Lumberjack
+
+###### Lumberjack
+
+The following services are required:
+
+- [NATS] - requires a [NATS] server to be running
+- [Elastic] - log destination
+
+A sample [.env](https://github.com/frmscoe/lumberjack/blob/ce6cda49bade1a0287bc0c603330e5fbb8159455/.env.example) is provided.
+
+| Variable Name         | Purpose                               | Example             |
+|-----------------------|---------------------------------------|---------------------|
+| NATS_SERVER           | Specifies the NATS server address      | `nats://localhost:4222` |
+| NATS_SUBJECT          | Defines the NATS subject/topic name    | `Lumberjack`        |
+| ELASTIC_SEARCH_VERSION| Specifies the Elasticsearch version   | `8.11`              |
+| ELASTIC_HOST          | Specifies the Elasticsearch host       | `http://localhost:9200` |
+| ELASTIC_USERNAME      | Username for Elasticsearch authentication | `elastic_user`   |
+| ELASTIC_PASSWORD      | Password for Elasticsearch authentication | `secretpassword` |
+| FLUSHBYTES            | Specifies the number of bytes to flush | `1024`              |
+
+
+###### Event-Sidecar
+A sample [.env](https://github.com/frmscoe/event-sidecar/blob/feb5e53b0801f60fa746e1720349cfb8c09e6c2b/.env.example) is provided.
+
+| Variable Name  | Purpose                              | Example                    |
+|----------------|--------------------------------------|----------------------------|
+| PORT           | Specifies the port number to use     | `8080`                     |
+| NATS_SERVER    | Specifies the NATS server address    | `nats://localhost:4222`    |
+| NATS_SUBJECT   | Defines the NATS subject/topic name  | `Lumberjack`               |
+
+> [!CAUTION]
+> Ensure the sidecar and lumberjack are pointing to the same instance of [NATS] and have the same subject
+
+###### Processor
+In your processor, add the [frms-coe-lib] as a dependency:
+
+```sh
+npm i @frmscoe/frms-coe-lib
+```
+
+In your processor, create an instance of a `LoggerService` and specify the [event-sidecar] address:
+
+```ts
+import { LoggerService } from '@frmscoe/frms-coe-lib';
+
+const logger = new LoggerService("localhost:8080");
+```
+> [!CAUTION]
+> Ensure the sidecar AND [lumberjack] start up before the processor so that logs are not missed.
```

Comment:
OK, I am missing what I would consider "tangible steps" to get a functional logging ecosystem up and running, at the level of what I would consider a "tutorial".

There are probably two perspectives for this kind of information here: first is from the perspective of a developer who wants to implement logging into their code, and second is from the perspective of an operator who wants to turn on (or dial up) the logging on their implementation. I imagine the steps for each of these would be somewhat different, but if I had to take a stab at the developer "tutorial" (on the assumption that the developer wants to spin up just the logging ecosystem from scratch so that they can implement logging in their processor):
1. Deploy NATS (for example, ref: https://github.com/frmscoe/docs/blob/05e5c292dcff908ec06825481cb99de2ecb31b74/Community/Tazama-Contribution-Guide.md#microprocessor-setup-instructions)
2. Deploy Elastic (no idea what this would look like, to be honest)
3. Deploy the side-car (steps)
4. Deploy Lumberjack (steps)
5. Update the processor (steps)

While some of the step groups are provided above, they are not necessarily in the logical sequence and with enough detail to guide the user.

From the perspective of an operator, I imagine the steps are a subset of the above, but with slightly more "production-focused" targets (NATS, Elastic) and these can be assumed to be deployed via Jenkins and referenced in this document (not sure about Elastic, but if it is "optional" in a production deployment, then the "option" should be exercisable in the EKS/AKS documentation through some kind of setting and the dependency on that setting should be referenced here in this documentation).

Reply:
I see. I will try to outline the steps according to the linked document as best as I can. Just to confirm:
> Deploy Elastic (no idea what this would look like, to be honest)
Deploy the side-car (steps)
Deploy Lumberjack (steps)

By "Deploy", are we mainly looking at getting the processor running (on a user's system)?